source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -167,8 +169,8 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
   // Try to write raw buffer when next call is done, even this is not in do[Read|Write] stack.
   if (raw_write_buffer_.length() > 0) {
     Network::IoResult result = raw_buffer_socket_->doWrite(raw_write_buffer_, false);
-    if (handshake_complete_ && raw_write_buffer_.length() > 0) {
-      write_buffer_contains_handshake_bytes_ = true;
+    if (handshake_complete_) {
```

Comment:
An error could be detected on doWrite which could cause the connection to be closed.  I think that this raiseEvent could then cause a crash if the connection is already closed.

Similarly, the call to doWrite after raiseEvent(Connected) needs to check the status of the connection. If the raiseEvent(Connected) closes the connection the doWrite may run into some trouble.  IIRC close on raiseEvent(Connected)  does happen in the case of SSL sockets.

Reply:
Which doWrite do you mean in "Similarly, the call to doWrite after raiseEvent(Connected) needs to check the status of the connection"? We do not call doWrite after raiseEvent in this PR.